src/storage/temporary_file_manager.cpp

```diff
@@ -180,113 +281,153 @@ void TemporaryFileHandle::RemoveTempBlockIndex(TemporaryFileLock &, idx_t index)
+	return *res.first->second;
+}
+
+void TemporaryFileMap::EraseFile(const TemporaryFileIdentifier &identifier) {
+	D_ASSERT(identifier.IsValid());
+	D_ASSERT(GetFile(identifier));
+	GetMapForSize(identifier.size).erase(identifier.file_index.GetIndex());
 }
 
 //===--------------------------------------------------------------------===//
-// TemporaryFileIndex
+// TemporaryFileCompressionLevel/TemporaryFileCompressionAdaptivity
 //===--------------------------------------------------------------------===//
+TemporaryFileCompressionAdaptivity::TemporaryFileCompressionAdaptivity() : last_uncompressed_write_ns(1) {
+	for (idx_t i = 0; i < LEVELS; i++) {
+		last_compressed_writes_ns[0] = 1;
+	}
+}
 
-TemporaryFileIndex::TemporaryFileIndex(idx_t file_index, idx_t block_index)
-    : file_index(file_index), block_index(block_index) {
+int64_t TemporaryFileCompressionAdaptivity::GetCurrentTimeNanos() {
+	return duration_cast<nanoseconds>(high_resolution_clock::now().time_since_epoch()).count();
 }
 
-bool TemporaryFileIndex::IsValid() const {
-	return block_index != DConstants::INVALID_INDEX;
+TemporaryCompressionLevel TemporaryFileCompressionAdaptivity::IndexToLevel(const idx_t index) {
+	switch (index) {
+	case 0:
+		return TemporaryCompressionLevel::ZSTD_MINUS_THREE;
+	case 1:
+		return TemporaryCompressionLevel::ZSTD_MINUS_ONE;
+	case 2:
+		return TemporaryCompressionLevel::ZSTD_ONE;
+	case 3:
+		return TemporaryCompressionLevel::ZSTD_THREE;
+	default:
+		throw InternalException("Unknown index in IndexToLevel");
+	}
 }
 
-//===--------------------------------------------------------------------===//
-// TemporaryFileManager
-//===--------------------------------------------------------------------===//
+idx_t TemporaryFileCompressionAdaptivity::LevelToIndex(TemporaryCompressionLevel level) {
+	switch (level) {
+	case TemporaryCompressionLevel::ZSTD_MINUS_THREE:
+		return 0;
+	case TemporaryCompressionLevel::ZSTD_MINUS_ONE:
+		return 1;
+	case TemporaryCompressionLevel::ZSTD_ONE:
+		return 2;
+	case TemporaryCompressionLevel::ZSTD_THREE:
+		return 3;
+	default:
+		throw InternalException("Unknown TemporaryFileCompressionLevel in LevelToIndex");
+	}
+}
 
-static idx_t GetDefaultMax(const string &path) {
-	D_ASSERT(!path.empty());
-	auto disk_space = FileSystem::GetAvailableDiskSpace(path);
-	// Use the available disk space
-	// We have made sure that the file exists before we call this, it shouldn't fail
-	if (!disk_space.IsValid()) {
-		// But if it does (i.e because the system call is not implemented)
-		// we don't cap the available swap space
-		return DConstants::INVALID_INDEX - 1;
+TemporaryCompressionLevel TemporaryFileCompressionAdaptivity::GetCompressionLevel() {
+	idx_t min_compression_idx = 0;
+	auto min_compressed_time = last_compressed_writes_ns[min_compression_idx].load();
+	for (idx_t compression_idx = 0; compression_idx < LEVELS; compression_idx++) {
+		const auto time = last_compressed_writes_ns[compression_idx].load();
```

Comment:
Maybe we want to keep these atomics per-thread as well similar to the memory usage and then flush them out periodically to reduce contention